stratum/hal/lib/barefoot/bf_sde_wrapper.cc

```diff
@@ -2208,6 +2208,172 @@ ::util::Status BfSdeWrapper::ReadRegisters(
+    for (size_t i = 0; i < table_size; ++i) {
+      // Meter key: $METER_INDEX
+      RETURN_IF_ERROR(SetField(table_key.get(), kMeterIndex, i));
+      RETURN_IF_BFRT_ERROR(table->tableEntryMod(
+          *real_session->bfrt_session_, bf_dev_tgt, *table_key, *table_data));
+    }
+  }
+
+  return ::util::OkStatus();
+}
+
+::util::Status BfSdeWrapper::ReadIndirectMeters(
+    int device, std::shared_ptr<BfSdeInterface::SessionInterface> session,
+    uint32 table_id, absl::optional<uint32> meter_index,
+    std::vector<uint32>* meter_indices, std::vector<uint64>* cirs,
+    std::vector<uint64>* cbursts, std::vector<uint64>* pirs,
+    std::vector<uint64>* pbursts, std::vector<bool>* in_pps) {
+  CHECK_RETURN_IF_FALSE(meter_indices);
+  CHECK_RETURN_IF_FALSE(cirs);
+  CHECK_RETURN_IF_FALSE(cbursts);
+  CHECK_RETURN_IF_FALSE(pirs);
+  CHECK_RETURN_IF_FALSE(pbursts);
+  ::absl::ReaderMutexLock l(&data_lock_);
+  auto real_session = std::dynamic_pointer_cast<Session>(session);
+  CHECK_RETURN_IF_FALSE(real_session);
+
+  auto bf_dev_tgt = GetDeviceTarget(device);
+  const bfrt::BfRtTable* table;
+  RETURN_IF_BFRT_ERROR(bfrt_info_->bfrtTableFromIdGet(table_id, &table));
+  std::vector<std::unique_ptr<bfrt::BfRtTableKey>> keys;
+  std::vector<std::unique_ptr<bfrt::BfRtTableData>> datums;
+
+  // Is this a wildcard read?
+  if (meter_index) {
+    keys.resize(1);
+    datums.resize(1);
+    RETURN_IF_BFRT_ERROR(table->keyAllocate(&keys[0]));
+    RETURN_IF_BFRT_ERROR(table->dataAllocate(&datums[0]));
+
+    // Key: $METER_INDEX
+    RETURN_IF_ERROR(SetField(keys[0].get(), kMeterIndex, meter_index.value()));
+    RETURN_IF_BFRT_ERROR(table->tableEntryGet(
+        *real_session->bfrt_session_, bf_dev_tgt, *keys[0],
+        bfrt::BfRtTable::BfRtTableGetFlag::GET_FROM_SW, datums[0].get()));
+  } else {
+    RETURN_IF_ERROR(GetAllEntries(real_session->bfrt_session_, bf_dev_tgt,
+                                  table, &keys, &datums));
+  }
+
+  meter_indices->resize(0);
+  cirs->resize(0);
+  cbursts->resize(0);
+  pirs->resize(0);
+  pbursts->resize(0);
+  in_pps->resize(0);
+  for (size_t i = 0; i < keys.size(); ++i) {
+    const std::unique_ptr<bfrt::BfRtTableData>& table_data = datums[i];
+    const std::unique_ptr<bfrt::BfRtTableKey>& table_key = keys[i];
+    // Key: $METER_INDEX
+    uint64 bf_meter_index;
+    RETURN_IF_ERROR(GetField(*table_key, kMeterIndex, &bf_meter_index));
+    meter_indices->push_back(bf_meter_index);
+
+    // Data: $METER_SPEC_*
+    std::vector<bf_rt_id_t> data_field_ids;
+    RETURN_IF_BFRT_ERROR(table->dataFieldIdListGet(&data_field_ids));
+    for (const auto& field_id : data_field_ids) {
+      std::string field_name;
+      RETURN_IF_BFRT_ERROR(table->dataFieldNameGet(field_id, &field_name));
+      if (field_name == kMeterCirKbps) {  // In kbits
+        uint64 cir;
+        RETURN_IF_BFRT_ERROR(table_data->getValue(field_id, &cir));
+        cirs->push_back(cir * 125);
```

Comment:
add comments like `Convert SDE kbits to P4RT bytes (* 1000 / 8).`

Reply:
Fixed by introducing helper functions.